# Финальный проект. Работа с SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.



***Задания***

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<a id='content'></a>
# СОДЕРЖАНИЕ

1. [Импорт библиотек. Подключение к БД. Просмотр таблиц.](#part1)

2. [Задание 1](#part2)

3. [Задание 2](#part3)

4. [Задание 3](#part4)

5. [Задание 4](#part5)

6. [Задание 5](#part6)

# Структура базы данных:

In [1]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2")

<a id='part1'></a>
## Импорт библиотек. Подключение к БД. Просмотр таблиц.

[*Назад к оглавлению*](#content)

Исследуйте таблицы — выведите первые строки;

In [2]:
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    sql = ''' SELECT * FROM {};'''.format(table)
    
    request = pd.io.sql.read_sql(sql, con = engine)
    
    print(f'Таблица - "{table}"')
    display(request.head(), request.info())

Таблица - "books"
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

None

Таблица - "authors"
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

None

Таблица - "publishers"
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

None

Таблица - "ratings"
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

None

Таблица - "reviews"
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

None

Таблица **`books`** содержит **данные о книгах**. Состоит из **1000 строк**, **6 столбцов**,  *пропуски отсутствуют*. 

Таблица **`authors`** содержит **данные об авторах**. Состоит из **636 строк**, **2 столбца**, *пропуски отсутствуют*. 

Таблица **`publishers`** содержит **данные об издательствах**. Состоит из **340 строк**, **2 столбца**, *пропуски отсутствуют*. 

Таблица **`ratings`** содержит **данные о пользовательских оценках книг**. Состоит из **6456 строк**, *пропуски отсутствуют*. 

Таблица **`reviews`**  содержит **данные о пользовательских обзорах на книги**. Состоит из **2793 строк**, **4 столбца**, *пропуски отсутствуют*. 

In [5]:
# Напишем функцию select, которая заворачивает пандосовкую функцию. 
# Нужно для того, чтобы каждый раз не писать всю эту авадакедавру, а писать красивенький селект

def select(sql):
    return pd.read_sql(sql,engine)

<a id='part2'></a>
## Задание №1

[*Назад к оглавлению*](#content)

***Посчитайте, сколько книг вышло после 1 января 2000 года***

In [6]:
sql1 = ''' SELECT count(book_id) books_cnt
FROM books b
WHERE publication_date > '2000-01-01'
'''

In [7]:
select(sql1)

books_cnt
0        819



После 1 января 2000 года вышло ***819 книг***

<a id='part3'></a>
## Задание №2

[*Назад к оглавлению*](#content)

***Для каждой книги посчитайте количество обзоров и среднюю оценку***

In [8]:
sql2 = ''' SELECT   b.book_id AS book_id,
                    b.title AS title,
                    review_cnt,
                    ROUND(average_rating,2) AS avg_rating
           FROM books b
           LEFT JOIN (SELECT book_id,
                             COUNT(review_id) AS review_cnt
                      FROM reviews   
                      GROUP BY 1) rv ON b.book_id = rv.book_id        
           LEFT JOIN (SELECT book_id,
                             AVG(rating) AS average_rating
                      FROM ratings     
                      GROUP BY 1) rt ON b.book_id = rt.book_id                  
'''

In [9]:
select(sql2)

book_id                                              title  review_cnt  \
0        652          The Body in the Library (Miss Marple  #3)         2.0   
1        273                                          Galápagos         2.0   
2         51                           A Tree Grows in Brooklyn         5.0   
3        951  Undaunted Courage: The Pioneering First Missio...         2.0   
4        839                                        The Prophet         4.0   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland         4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)         2.0   
997      148                                          Christine         3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999      828                           The Plot Against America         2.0   

     avg_rating  
0          4.50  
1          4.50  
2          4.25  
3          4.00  
4          4.29  
..          ...  
995        4.23  
996        5.00  
997        3.43  
998        3.50  
999        3.00  

[1000 rows x 4 columns]

<a id='part4'></a>
## Задание №3

[*Назад к оглавлению*](#content)

***Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры***

In [10]:
sql3 = ''' SELECT t.publisher_id,
                  t.publisher,
                  t.n_books
           FROM (SELECT b.publisher_id AS publisher_id,
                        COUNT(b.book_id) AS n_books,
                        p.publisher AS publisher 
                 FROM books b
                 LEFT JOIN publishers p ON b.publisher_id=p.publisher_id
                 WHERE num_pages>50
                 GROUP BY 1, 3
                 ORDER BY 2 DESC) t
           LIMIT 1   
'''

In [11]:
select(sql3)

publisher_id      publisher  n_books
0           212  Penguin Books       42



***Издательство***, выпустившее наибольшее число книг(за исключением брошюр), ***называется Penguin Books и выпустило 42 книги.***

<a id='part4'></a>
## Задание №4

[*Назад к оглавлению*](#content)

***Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.***

In [12]:
sql4 = ''' SELECT b.author_id AS author_id,
                  a.author AS author,
                  ROUND(AVG(t.book_rating_avg),3) AS author_rating_avg  
           FROM books b
           LEFT JOIN authors a ON a.author_id = b.author_id         
           INNER JOIN (SELECT book_id,
                              AVG(rating) AS book_rating_avg
                       FROM ratings
                       GROUP BY book_id
                       HAVING COUNT(*) >= 50) t on t.book_id = b.book_id     
           GROUP BY 1, 2                   
           ORDER BY 3 DESC                
           LIMIT 1                     
                      
'''

In [13]:
select(sql4)

author_id                      author  author_rating_avg
0        236  J.K. Rowling/Mary GrandPré              4.284

 

Автор с самой высокой средней оценкой книг - ***Джоан Роулинг (средний рейтинг - 4.28)***. Не удивительно =)

<a id='part4'></a>
## Задание №5

[*Назад к оглавлению*](#content)

***Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок***

In [14]:
sql5 = ''' SELECT ROUND(AVG(cnt_reviews)) AS avg_reviews  
           FROM (SELECT rv.username,
                        COUNT(text) AS cnt_reviews
                 FROM reviews rv
                 JOIN (SELECT username,
                              COUNT(rating)
                       FROM ratings
                       GROUP BY 1
                       HAVING COUNT(rating) > 50) t ON t.username=rv.username
                 GROUP BY 1)  f         
                      
'''

In [15]:
select(sql5)

avg_reviews
0         24.0



***Среднее количество обзоров*** от пользователей, которые поставили больше 50 оценок ***равно 24***